src/map/status.c

```diff
@@ -12278,38 +12279,118 @@ bool status_readdb_sizefix(char* fields[], int columns, int current)
+		bool duplicate[MAX_REFINE];
+		int bonus_[MAX_REFINE], rnd_bonus[MAX_REFINE], chance_[MAX_REFINE];
+
+		memset(&duplicate,0,sizeof(duplicate));
+		memset(&bonus_,0,sizeof(bonus_));
+		memset(&rnd_bonus,0,sizeof(rnd_bonus));
+		memset(&chance_,0,sizeof(chance_));
+		i=0;
+		while( (t = libconfig->setting_get_elem(rate,i++)) && config_setting_is_group(t) ) {
+			int level=0, chance=0, bonus=0;
+			char *rlvl = config_setting_name(t);
+			
+			if(!strspn(&rlvl[strlen(rlvl)-1], "0123456789") || (level = atoi(strtok(rlvl, "Lv"))) <= 0 ) {
+				ShowError("status_readdb_refine_libconfig_sub: Invalid refine level format '%s' for entry %s in \"%s\".\n", rlvl, name, source);
+				continue;
+			} else if (level <= 0 || level > MAX_REFINE ) {
+				ShowError("status_readdb_refine_libconfig_sub: Out of range refine level '%s' for entry %s in \"%s\".\n", rlvl, name, source);
+				continue;
+			} else if( duplicate[level-1] ) {
+				ShowWarning("status_readdb_refine_libconfig_sub: duplicate rate '%s' for entry %s in \"%s\".\n", rlvl, name, source);
+			} else duplicate[level-1] = true;
+			level--;
+			chance_[level] = libconfig->setting_lookup_int(t, "Chance", &chance)?chance:100;
+			bonus_[level] = bonus_per_level + libconfig->setting_lookup_int(t, "Bonus", &bonus)?bonus:0;
+			if ( level >= rnd_bonus_lv - 1 ) rnd_bonus[level] = rnd_bonus_v * (level - rnd_bonus_lv + 2);
+		}
+		for(i=0; i < MAX_REFINE; i++) {
+			status->dbs->refine_info[type].chance[i] = chance_[i]?chance_[i]:100;
+			status->dbs->refine_info[type].randombonus_max[i] = rnd_bonus[i];
+			bonus_[i] += i?bonus_[i-1]:0;
+			status->dbs->refine_info[type].bonus[i] = bonus_[i]?bonus_[i]:0;
+		}
+	} else {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing refine rates for entry '%s' in \"%s\", skipping.\n", name, source);
+		return 0;
+	}
+	return type+1;
+}
 
-		status->dbs->refine_info[current].bonus[i] = bonus_per_level + atoi(delim+1);
-		if (i > 0)
-			status->dbs->refine_info[current].bonus[i] += status->dbs->refine_info[current].bonus[i-1];
+/**
+ * Reads from a libconfig-formatted refine_db.conf file.
+ *
+ * @param *filename File name, relative to the database path.
+ * @return The number of found entries.
+ */
+int status_readdb_refine_libconfig(const char *filename) {
+	bool duplicate[REFINE_TYPE_MAX];
+	config_t refine_db_conf;
+	config_setting_t *r;
+	char filepath[256];
+	int i = 0, count = 0,type = 0;
+	
+	sprintf(filepath, "%s/%s", map->db_path, filename);
+	memset(&duplicate,0,sizeof(duplicate));
+	if( libconfig->read_file(&refine_db_conf, filepath) ) {
+		ShowError("can't read %s\n", filepath);
+		return 0;
 	}
-	return true;
+	
+	while( (r = libconfig->setting_get_elem(refine_db_conf.root,i++)) ) {
+		char *name = config_setting_name(r);
+		if((type=status->readdb_refine_libconfig_sub(r, name, filename))) {
+			if( duplicate[type-1] ) {
+				ShowError("status_readdb_refine_libconfig: duplicate entry for %s in \"%s\", skipping... \n", name, filename);
```

Comment:
The message mentions skipping, but the code isn't skipping it. `status_readdb_refine_libconfig_sub` overwrites any existing values, rather than skipping. Skipping might be a little tricky (you could move the code that retrieves the type here if you really want to skip it).